cmd/boostd/genindex.go

```diff
@@ -0,0 +1,221 @@
+package main
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	_ "net/http/pprof"
+	"os"
+	"time"
+
+	"github.com/filecoin-project/boost/cmd/lib"
+	"github.com/filecoin-project/boost/piecedirectory"
+	"github.com/filecoin-project/boostd-data/model"
+	lcli "github.com/filecoin-project/lotus/cli"
+	"github.com/ipfs/go-cid"
+	carv1 "github.com/ipld/go-car"
+	carv2 "github.com/ipld/go-car/v2"
+	"github.com/multiformats/go-varint"
+	"github.com/urfave/cli/v2"
+)
+
+var genindexCmd = &cli.Command{
+	Name:  "genindex",
+	Usage: "Generate index for a given piececid and store in the piece directory",
+	Flags: []cli.Flag{
+		&cli.IntFlag{
+			Name:  "add-index-throttle",
+			Usage: "the maximum number of add index operations that can run in parallel",
+			Value: 4,
+		},
+		&cli.StringFlag{
+			Name:     "api-fullnode",
+			Usage:    "the endpoint for the full node API",
+			Required: false,
+		},
+		&cli.StringFlag{
+			Name:     "api-storage",
+			Usage:    "the endpoint for the storage node API",
+			Required: false,
+		},
+		&cli.StringFlag{
+			Name:     "api-piece-directory",
+			Usage:    "the endpoint for the piece directory API",
+			Required: false,
+		},
+		&cli.StringFlag{
+			Name:     "piececid",
+			Usage:    "piececid to index",
+			Required: false,
+		},
+		&cli.StringFlag{
+			Name:     "filepath",
+			Usage:    "the path to the car",
+			Required: false,
+		},
+	},
+	Action: func(cctx *cli.Context) error {
+		ctx := lcli.ReqContext(cctx)
+
+		// parse piececid
+		piececid, err := cid.Decode(cctx.String("piececid"))
+		if err != nil {
+			return err
+		}
+		fmt.Println("piececid to index: ", piececid)
+
+		// connect to the piece directory service
+		pdClient := piecedirectory.NewStore()
+		defer pdClient.Close(ctx)
+		err = pdClient.Dial(ctx, cctx.String("api-piece-directory"))
+		if err != nil {
+			return fmt.Errorf("error while connecting to piece directory service: %w", err)
+		}
+
+		// connect to the full node
+		fnApiInfo := cctx.String("api-fullnode")
+		fullnodeApi, ncloser, err := lib.GetFullNodeApi(ctx, fnApiInfo, log)
+		if err != nil {
+			return fmt.Errorf("getting full node API: %w", err)
+		}
+		defer ncloser()
+
+		// connect to the storage API and create a sector accessor
+		storageApiInfo := cctx.String("api-storage")
+		sa, storageCloser, err := lib.CreateSectorAccessor(ctx, storageApiInfo, fullnodeApi, log)
+		if err != nil {
+			return err
+		}
+		defer storageCloser()
+
+		pr := &piecedirectory.SectorAccessorAsPieceReader{SectorAccessor: sa}
+
+		pd := piecedirectory.NewPieceDirectory(pdClient, pr, cctx.Int("add-index-throttle"))
+
+		var r *carv2.Reader
+		var headerDataSize uint64
+
+		if filepath := cctx.String("filepath"); filepath != "" {
+			fmt.Println("fetching piececid from car file: ", filepath)
+
+			r, err = carv2.OpenReader(filepath)
+			if err != nil {
+				return fmt.Errorf("open reader: %w", err)
+			}
+			defer r.Close()
+
+			if r.Version == 1 {
```

Comment:
Typically we deal with carV1 but if user supplies a carv2 we should accept it as well and directly take the indexed from it. WDYT?
And we don't want to do that then perhaps we should let the user know that they have supplied a cav2 and return from there.

Reply:
> Typically we deal with carV1 but if user supplies a carv2 we should accept it as well and directly take the indexed from it.

I think we have agreed that we always re-generate indices and don't take them from the client even if they are there.